src/validation.cpp

```diff
@@ -2782,14 +2781,12 @@ static bool NotifyHeaderTip(CChainState& chainstate) LOCKS_EXCLUDED(cs_main) {
 
         if (pindexHeader != pindexHeaderOld) {
             fNotify = true;
-            assert(std::addressof(::ChainstateActive()) == std::addressof(chainstate));
-            fInitialBlockDownload = chainstate.IsInitialBlockDownload();
```

Comment:
In commit "validation: make NotifyHeaderTip not rely on CChainState object" (2369045c6909bc242b996976d635cd7e6f9c7f62)

I know this commit came from another reviewer's suggestion, so feel free to ignore this comment, but this commit doesn't seem great to me. It's not simplifying code or semantics, and it seems like it might send a less accurate notification if it is unlocking and relocking `cs_main` between getting `pindexHeader` and `is_in_ibd` values instead of getting both values under the same lock.

Reply:
I tend to agree. What's the benefit of this commit?

I haven't looked in great detail, but I imagine we probably want `pindexBestHeader` to move from being a global to being a member of CChainState, guarded by cs_main.

Reply:
As far as I can tell, this commit is necessary if we both (i) want chainstate usage to be covered by `cs_main` and (ii) want to preserve the lock exclusion annotation on `NotifyHeaderTip`; there is no way to satisfy both while passing in a chainstate as a parameter.

That said, I have no problem peeling off the last two commits and deferring locking considerations for another PR.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19806#discussion_r571293050

> As far as I can tell, this commit is necessary if we both (i) want chainstate usage to be covered by `cs_main` and (ii) want to preserve the lock exclusion annotation on `NotifyHeaderTip`; there is no way to satisfy both while passing in a chainstate as a parameter.
> 
> That said, I have no problem peeling off the last two commits and deferring locking considerations for another PR.

Unless there's a specific reason the commits are needed here, I'd peel them off. The original reason to involve locking in this PR according to https://github.com/bitcoin/bitcoin/pull/19806#discussion_r561023961 was to guard `m_active_chainstate`, because at the time it was unguarded, and this PR introduced code changing the `m_active_chainstate` value. But now that `m_active_chainstate` is guarded on master (https://github.com/bitcoin/bitcoin/pull/19806#pullrequestreview-584279133), I think these commits only complicate things here and would make more sense as a separate PR.

On https://github.com/bitcoin/bitcoin/commit/2369045c6909bc242b996976d635cd7e6f9c7f62, I don't think there's a problem continuing to use chainstate inside NotifyHeaderTip regardless of the LOCKS_EXCLUDED annotation, because NotifyHeaderTip is still locking cs_main internally a few lines up above this on line 2779